src/schemes/boundary/open_boundary/boundary_zones.jl

```diff
@@ -130,9 +153,118 @@ function BoundaryZone(; plane, plane_normal, density, particle_spacing,
+                            "a vector holding the density of each particle, or a scalar"))
+    else
+        if reference_density isa Function
+            test_result = reference_density(zeros(NDIMS), 0.0)
+            if length(test_result) != 1
+                throw(ArgumentError("`reference_density` function must be a scalar function"))
+            end
+        end
+        # We need this dummy for type stability reasons
+        density_dummy = ELTYPE(Inf)
+        density_ref = wrap_reference_function(reference_density, density_dummy)
+    end
+
+    prescribed_pressure = isnothing(reference_pressure) ? false : true
+    prescribed_density = isnothing(reference_density) ? false : true
+    prescribed_velocity = isnothing(reference_velocity) ? false : true
+
+    reference_values = (reference_velocity=velocity_ref, reference_pressure=pressure_ref,
+                        reference_density=density_ref)
+
+    coordinates_svector = reinterpret(reshape, SVector{NDIMS, ELTYPE}, ic.coordinates)
+
+    if prescribed_pressure
+        ic.pressure .= pressure_ref.(coordinates_svector, 0)
+    end
+    if prescribed_density
+        ic.density .= density_ref.(coordinates_svector, 0)
+        ic.mass .= ic.density * ic.particle_spacing^NDIMS
+    end
+    if prescribed_velocity
+        ic.velocity .= stack(velocity_ref.(coordinates_svector, 0))
+    end
+
     return BoundaryZone(ic, spanning_set_, zone_origin, zone_width,
-                        flow_direction, plane_normal_, boundary_type,
-                        average_inflow_velocity)
+                        flow_direction, plane_normal_, reference_values,
+                        average_inflow_velocity, prescribed_density, prescribed_pressure,
+                        prescribed_velocity)
+end
+
+function boundary_type_name(boundary_zone::BoundaryZone)
+    (; flow_direction, plane_normal) = boundary_zone
+
+    # Bidirectional flow
+    isnothing(flow_direction) && return "bidirectional_flow"
+
+    # Outflow
+    signbit(dot(flow_direction, plane_normal)) && return "outflow"
+
+    # Inflow
+    return "inflow"
+end
+
+function Base.show(io::IO, boundary_zone::BoundaryZone)
+    @nospecialize boundary_zone # reduce precompilation time
+
+    print(io, "BoundaryZone(")
+    print(io, ") with ", nparticles(boundary_zone.initial_condition), " particles")
+end
+
+function Base.show(io::IO, ::MIME"text/plain", boundary_zone::BoundaryZone)
+    @nospecialize boundary_zone # reduce precompilation time
+
+    if get(io, :compact, false)
+        show(io, boundary_zone)
+    else
+        summary_header(io, "BoundaryZone")
+        summary_line(io, "boundary type", boundary_type_name(boundary_zone))
+        summary_line(io, "#particles", nparticles(boundary_zone.initial_condition))
+        summary_line(io, "width", round(boundary_zone.zone_width, digits=3))
+        summary_footer(io)
+    end
 end
 
 function set_up_boundary_zone(plane, plane_normal, flow_direction, density,
```

Comment:
1. Why does this function have a kwarg? It is only called once and only used internally.
2. Why don't you include `flow_direction` in this function as well? Then you can clean up the constructor a bit.